react-ui/src/pages/analyzeVotes/PearsonsCorrelation.jsx

```diff
@@ -0,0 +1,133 @@
+/* eslint-disable max-len */
+/* eslint-disable react/forbid-prop-types */
+import PropTypes from 'prop-types';
+import { useCallback, useEffect, useMemo } from 'react';
+import Plot from 'react-plotly.js';
+
+import {
+  roundTwoDecimals, createTraces, trimUsername,
+} from './functions';
+import MARKERS from './markers';
+
+const GROUP = { selected: 0, none: 1, few: 2 };
+
+/**
+ * Compare user similarity
+ */
+function PearsonsCorrelation({
+  username, username2, votes, entryPositionLookup, setUsername2, voteMinimum,
+}) {
+  /**
+   * CALCULATE PEARSONS
+   */
+  const pearsons = useMemo(() => {
+    // USER DOES NOT EXISTS
+    if (!username) { return []; }
+
+    const numberOfEntries = Object.keys(entryPositionLookup).length;
+    const allVotesByUser = votes.reduce((acc, vote) => {
+      if (!acc[vote.username]) {
+        acc[vote.username] = new Array(numberOfEntries);
+      }
+      acc[vote.username][entryPositionLookup[vote.entryId]] = vote.rating;
+      return acc;
+    }, {});
+
+    const userVotes = allVotesByUser[username];
+    // USER DID NOT VOTE
+    if (!userVotes) { return []; }
+
+    return Object.keys(allVotesByUser).filter((u) => u !== username).map((u) => {
+      const otherUser = allVotesByUser[u];
+      let sum1 = 0; let sum2 = 0; let sum1Sq = 0; let sum2Sq = 0; let sumProduct = 0;
+      let n = 0;
+      let numVotes = 0;
+      otherUser.forEach((vote, index) => {
+        const userVote = userVotes[index];
+        if (vote > -1 && userVote > -1) {
+          sum1 += vote;
+          sum2 += userVote;
+          sum1Sq += vote * vote;
+          sum2Sq += userVote * userVote;
+          sumProduct += vote * userVote;
+          n += 1;
+        }
+        if (vote > -1) { numVotes += 1; }
+      });
+      const num = sumProduct - ((sum1 * sum2) / n);
+      const den = Math.sqrt((sum1Sq - ((sum1 * sum1) / n)) * (sum2Sq - ((sum2 * sum2) / n)));
+
+      if (n === 0 || den === 0) { return { username: u, correlation: undefined }; }
+
+      return { username: u, correlation: num / den, group: numVotes < voteMinimum ? GROUP.few : undefined };
+    }).filter((p) => p.correlation !== undefined)
```

Comment:
Instead of
```js
const result = arr.map((v) => cond ? v + 1 : undefined).filter((v) => v !== undefined);
```
you can just use a `reduce`:
```js
const result = arr.reduce((acc, cur) => cond ? [...acc, v + 1] : acc);
```
Or, if it's clearer:
```js
const result = [];
arr.forEach((v) => if (cond) { result.push(v + 1); });
```

Reply:
I went with the latter, thanks!